guardian/typerighter
# Add publish history

## What does this change?

Adds a publish history to our rules, and displays it in the UI.

|Nothing published yet|Published|Subsequent revisions|
|-|-|-|
|<img width="384" alt="Screenshot 2023-05-24 at 11 46 31" src="https://github.com/guardian/typerighter/assets/7767575/1e5ed4ee-9b5f-4e5e-a5cf-0fc411b5e131">|<img width="387" alt="Screenshot 2023-05-24 at 11 48 25" src="https://github.com/guardian/typerighter/assets/7767575/e34470e0-094a-4918-841e-e003c98bf668">|<img width="390" alt="Screenshot 2023-05-23 at 17 41 18" src="https://github.com/guardian/typerighter/assets/7767575/f00e9712-0caf-4fba-8fee-6360ed6d2ad2">|

We now keep every live rule we ever write, but ensure that only one is active at a given time.

In the database, we model this by removing the unique constraint on external ids for our live table, and adding a constraint for (external id, revisionId), to ensure that it's not possible to write the same rule and revision multiple times. We then add another property to the live table, `is_active`, with a unique constraint on `is_active` that only applies when `is_active` is true, to prevent more than one live rule being active at a time.

Our `rule/:id` endpoint then serves both the current draft rule and the rule history on every read.

On the frontend, I've refactored a bit to move the 'create rule' button out of the rule form itself, and abstracted the individual rule fetch logic into its own hook, to separate the concerns of fetching, managing the form open state, and the form itself.

As a result, this PR is a bit larger than expected! The server-side and client-side commits should be well separated, but I'm happy to split this into two PRs if it'd be easier to review. One advantage of keeping just one PR – it's easier to verify that the publish logic is working as expected.

## How to test

The automated tests that have been adjusted for publication and the active state, and the new tests for getting rule data, should pass.

We don't yet have a publish button – this will come with its own server-side validation logic, so I thought given the size of this PR it'd be better to ship that separately. In its absence, publish rules per the guidance in #324:

- Publish a rule successfully, and the live version should appear in the rule history
- Attempting to publish a rule again should result in an error (you cannot publish the same rule twice)
- Editing that rule to give it a new revisionId should allow it to be published again. The new version should appear in the publication history.
- Examining the history of the rules pulled in from the sheet by clicking the 'refresh' button should show that they've been created by the sheet (per the second screenshot above)
